### ISSUE TITLE
OPC UA Generating new self signed certificates loop

### ISSUE THREAD

**Issue by Jayclifford345** [influxdata/telegraf#11454]
### Relevant telegraf.conf

```toml
# Telegraf Configuration
#
# Telegraf is entirely plugin driven. All metrics are gathered from the
# declared inputs, and sent to the declared outputs.
#
# Plugins must be declared in here to be active.
# To deactivate a plugin, comment out the name and any variables.
#
# Use 'telegraf -config telegraf.conf -test' to see what metrics a config
# file would generate.
#
# Environment variables can be used anywhere in this config file, simply surround
# them with ${}. For strings the variable must be within quotes (ie, "${STR_VAR}"),
# for numbers and booleans they should be plain (ie, ${INT_VAR}, ${BOOL_VAR})


# Global tags can be specified here in key="value" format.
[global_tags]
  # dc = "us-east-1" # will tag all metrics with dc=us-east-1
  # rack = "1a"
  ## Environment variables can be used as tags, and throughout the config file
  # user = "$USER"


# Configuration for telegraf agent
[agent]
  ## Default data collection interval for all inputs
  interval = "5s"
  ## Rounds collection interval to 'interval'
  ## ie, if interval="10s" then always collect on :00, :10, :20, etc.
  round_interval = true

  ## Telegraf will send metrics to outputs in batches of at most
  ## metric_batch_size metrics.
  ## This controls the size of writes that Telegraf sends to output plugins.
  metric_batch_size = 1000

  ## Maximum number of unwritten metrics per output.  Increasing this value
  ## allows for longer periods of output downtime without dropping metrics at the
  ## cost of higher maximum memory usage.
  metric_buffer_limit = 10000

  ## Collection jitter is used to jitter the collection by a random amount.
  ## Each plugin will sleep for a random time within jitter before collecting.
  ## This can be used to avoid many plugins querying things like sysfs at the
  ## same time, which can have a measurable effect on the system.
  collection_jitter = "5s"

  ## Default flushing interval for all outputs. Maximum flush_interval will be
  ## flush_interval + flush_jitter
  flush_interval = "10s"
  ## Jitter the flush interval by a random amount. This is primarily to avoid
  ## large write spikes for users running a large number of telegraf instances.
  ## ie, a jitter of 5s and interval 10s means flushes will happen every 10-15s
  flush_jitter = "5s"

  ## By default or when set to "0s", precision will be set to the same
  ## timestamp order as the collection interval, with the maximum being 1s.
  ##   ie, when interval = "10s", precision will be "1s"
  ##       when interval = "250ms", precision will be "1ms"
  ## Precision will NOT be used for service inputs. It is up to each individual
  ## service input to set the timestamp at the appropriate precision.
  ## Valid time units are "ns", "us" (or "µs"), "ms", "s".
  precision = ""


  ## Override default hostname, if empty use os.Hostname()
  hostname = ""
  ## If set to true, do no set the "host" tag in the telegraf agent.
  omit_hostname = false

  debug = true

  quiet = false


###############################################################################
#                            OUTPUT PLUGINS                                   #
###############################################################################

# Configuration for sending metrics to InfluxDB 2.0
[[outputs.influxdb_v2]]
alias = "opc-ua-db"
  ## The URLs of the InfluxDB cluster nodes.
  ##
  ## Multiple URLs can be specified for a single cluster, only ONE of the
  ## urls will be written to each interval.
  ##   ex: urls = ["https://us-west-2-1.aws.cloud2.influxdata.com"]
  urls = ["${INFLUX_HOST}"]

  ## Token for authentication.
  token = "${INFLUX_TOKEN}"

  ## Organization is the name of the organization you wish to write to.
  organization = "${INFLUX_ORG}"

  ## Destination bucket to write into.
  bucket = "${INFLUX_BUCKET}"



###############################################################################
#                            INPUT PLUGINS                                    #
###############################################################################
[[inputs.opcua]]
  ## Metric name
  name = "opcua"
  #
  ## OPC UA Endpoint URL
  endpoint = "opc.tcp://localhost:49320"
  #
  ## Maximum time allowed to establish a connect to the endpoint.
  # connect_timeout = "10s"
  #
  ## Maximum time allowed for a request over the estabilished connection.
  # request_timeout = "5s"
  #
  ## Security policy, one of "None", "Basic128Rsa15", "Basic256",
  ## "Basic256Sha256", or "auto"
   security_policy = "auto"
  #
  ## Security mode, one of "None", "Sign", "SignAndEncrypt", or "auto"
   security_mode = "auto"
  #
  ## Path to cert.pem. Required when security mode or policy isn't "None".
  ## If cert path is not supplied, self-signed cert and key will be generated.
   certificate = ""
  #
  ## Path to private key.pem. Required when security mode or policy isn't "None".
  ## If key path is not supplied, self-signed cert and key will be generated.
   private_key = ""
  #
  ## Authentication Method, one of "Certificate", "UserName", or "Anonymous".  To
  ## authenticate using a specific ID, select 'Certificate' or 'UserName'
   auth_method = "UserName"
  #
  ## Username. Required for auth_method = "UserName"
   username = "telegraf"
  #
  ## Password. Required for auth_method = "UserName"
   password = "telegraftelegraf"
  #
  ## Option to select the metric timestamp to use. Valid options are:
  ##     "gather" -- uses the time of receiving the data in telegraf
  ##     "server" -- uses the timestamp provided by the server
  ##     "source" -- uses the timestamp provided by the source
  # timestamp = "gather"
  #
  ## Node ID configuration
  ## name              - field name to use in the output
  ## namespace         - OPC UA namespace of the node (integer value 0 thru 3)
  ## identifier_type   - OPC UA ID type (s=string, i=numeric, g=guid, b=opaque)
  ## identifier        - OPC UA ID (tag as shown in opcua browser)
  ## tags              - extra tags to be added to the output metric (optional)
  ## Example:
  ## {name="ProductUri", namespace="0", identifier_type="i", identifier="2262", tags=[["tag1","value1"],["tag2","value2]]}
   nodes = [
    {name="Current", namespace="2", identifier_type="s", identifier="Bender.Bender.Current"},
    {name="Bad_Quantity", namespace="2", identifier_type="s", identifier="Bender.Bender.Bad_Quantity"},
  ]
  #
  ## Node Group
  ## Sets defaults for OPC UA namespace and ID type so they aren't required in
  ## every node.  A group can also have a metric name that overrides the main
  ## plugin metric name.
  ##
  ## Multiple node groups are allowed
  #[[inputs.opcua.group]]
  ## Group Metric name. Overrides the top level name.  If unset, the
  ## top level name is used.
  # name =
  #
  ## Group default namespace. If a node in the group doesn't set its
  ## namespace, this is used.
  # namespace =
  #
  ## Group default identifier type. If a node in the group doesn't set its
  ## namespace, this is used.
  # identifier_type =
  #
  ## Node ID Configuration.  Array of nodes with the same settings as above.
  # nodes = [
  #  {name="", namespace="", identifier_type="", identifier=""},
  #  {name="", namespace="", identifier_type="", identifier=""},
  #]
```

### Logs from Telegraf

```text
2022-07-05T12:09:18Z I! Starting Telegraf 1.23.0
2022-07-05T12:09:18Z I! Loaded inputs: opcua
2022-07-05T12:09:18Z I! Loaded aggregators: 
2022-07-05T12:09:18Z I! Loaded processors: 
2022-07-05T12:09:18Z I! Loaded outputs: influxdb_v2
2022-07-05T12:09:18Z I! Tags enabled: host=Jays-MBP
2022-07-05T12:09:18Z I! [agent] Config: Interval:5s, Quiet:false, Hostname:"Jays-MBP", Flush Interval:10s
2022-07-05T12:09:18Z D! [agent] Initializing plugins
2022-07-05T12:09:18Z D! [agent] Connecting outputs
2022-07-05T12:09:18Z D! [agent] Attempting connection to [outputs.influxdb_v2::opc-ua-db]
2022-07-05T12:09:18Z D! [agent] Successfully connected to outputs.influxdb_v2::opc-ua-db
2022-07-05T12:09:18Z D! [agent] Starting service inputs
2022-07-05T12:09:23Z E! [inputs.opcua] Error in plugin: error in Client Connection: EOF
2022-07-05T12:09:29Z E! [inputs.opcua] Error in plugin: error in Client Connection: EOF
2022-07-05T12:09:30Z D! [outputs.influxdb_v2::opc-ua-db] Buffer fullness: 0 / 10000 metrics
```

### System info

MacOS, Linux, Windows

### Steps to reproduce

Hopefully, I can explain it in steps:
Telegraf auto generates a self-signed certificate and sends this to the OPC UA server.

1. Telegraf then proceeds to request tags but fails since the OPC UA server has rejected the client because the certificates are untrusted.
2. User accepts the certificates within their OPC UA server. Telegraf continues to fall over with the same error message.
3. On telegraf restart a new certificate and key is generated and requires authorisation from the server once again.
4. Some times you get lucky and catch the trust store quick enough and Telegraf will operate as expected. It's hard to gauge however as certificate errors output a EOF error.

### Expected behavior

Auto self-signed certificates are generated once and not recreated if they already exist. 

### Actual behavior

New certificates are generated on each run. 

### Additional info

Current workaround. Generate self-signed manually: https://github.com/InfluxCommunity/Telegraf-Community-Configs/tree/master/Kepware/certificates

**Comment by R290** [comment]
As far as I can see this is by design, so not necessarily a bug. On each call to `generateCert` a new, unique temporary directory is created for storing the newly generated certificates. We could change it to use the `certificate` and  `private_key` path options to store them (if they do not exist)? I'm not sure how this would work from a backwards compatability standpoint.

**Comment by Jayclifford345** [comment]
@R290, do you know where the temp directories are located? If users can easily get this directory to extract these auto-generated certificates they could put them in a more permanent location and manually assign the paths to them.

 I personally think that auto-generated certificates should be kept and reused unless deleted by the user. Otherwise, the user continuously has to approve new certificates on restart.

**Comment by R290** [comment]
The exact location depends on the OS. For Linux this would be `/tmp/ssc#` where `#` is a 10-digit random number. The best way might be to add an option `skip_certificate_creation` which should default to true. Asking the user to copy the certificates from some temporary directory is a bit cumbersome.

**Comment by Jayclifford345** [comment]
Agreed I think this would be a great catch-all. I will generate a feature request.

**Comment by srebhan** [comment]
@Jayclifford345 I'm not 100% sure I get your point. In case you don't specify a certificate a self-signed one is created _once_ at startup (just as the documentation says). The connection is attempted and will fail if the cert is not trusted, but as soon as it is, the connection (attempted in each gather cycle) will succeed.
If you want to persist the certificate, just specify it in the option. I don't see any "bug" so far... What is your intended behavior?

**Comment by Jayclifford345** [comment]
@srebhan Which option allows them to persist the certificate between Telegraf restarts? Currently based on the logic discussed the user would have to track down the temporary created certificate directory and assign these manually within their config. Will these temporary directories also persist after restart?  Perhaps this is more of a documentation issue. Though in its current form I think it's bad practice to continuously generate a new certificate on each restart of Telegraf when using the auto-generate feature.

**Comment by Jayclifford345** [comment]
Here is the video of the certificate loop: https://drive.google.com/file/d/1Y3EPuTyY1tciAHfpj-kTP3EqlgJ4-oJs/view?usp=sharing

At the end, it appears clearing the full certificate store seemed to get things to work again which was strange. Nonetheless, before that the results were undesirable.

**Comment by R290** [comment]
I've tracked down the origin of the EOF to https://github.com/gopcua/opcua/blob/bff20f0e5b6883be0cf4916290ed055e547f15dc/uasc/secure_channel.go#L690. This seems to be triggered when the secure channel is disconnected. Changing this would be part of the gopcua library and not telegraf itself.

It might be good to add a Wireshark capture to your video to see what's happening with establishing the secure connection in the background? It might be that the certificate store is only read when establishing the connection and therefore data can still go through when the connection is there.

**Comment by srebhan** [comment]
@Jayclifford345 the video shows it's working. :-P Just kidding... :-)

Regarding the `EOF` error, we have no reliable way to find out if failing certificates is the only case we get an `EOF`. However, we could add a error-log with a hint, e.g. `Connection closed, maybe your certificate isn't trusted?` or similar?

For certificate generation, I don't see where multiple self-signed certificates can be created in a single instance. The code flow is

1. [`Init()` calls `setupOptions()`](https://github.com/influxdata/telegraf/blob/8c0cf660c95b6b69abaab79d817749f735206465/plugins/inputs/opcua/opcua.go#L138)
2. [`setupOptions()` calls `generateCert()`](https://github.com/influxdata/telegraf/blob/8c0cf660c95b6b69abaab79d817749f735206465/plugins/inputs/opcua/opcua.go#L400) which then generates the certificate and writes it to a temporary location.
3. The returned certificate gets assigned to `o.Certificate`, preventing [`generateClientOpts()`](https://github.com/influxdata/telegraf/blob/8c0cf660c95b6b69abaab79d817749f735206465/plugins/inputs/opcua/opcua.go#L407) to generate another cert. _Note to myself_: Why do we do this twice?

**Comment by srebhan** [comment]
Honestly speaking, I think Telegraf should not generate _any_ certificate here! :-) We are talking about avoiding a **copy of a file**  and add an config option. I thought of changing the behavior to require the user to specify a certificate file and if that file does not exist to create a temporary, self-signed cert. However, this will mask configuration errors (e.g. a typo in the filename) and might cause hard-to-debug issues because you cannot directly see that the certificate is self-signed...

In the end we would need an option to _enable_ self-signed certificate generation, making things worse. As I said, I'd rather remove the "feature" (i.e. move it out of Telegraf) and give clear instructions on how to generate such a certificate (maybe with a script or similar).

**Comment by Jayclifford345** [comment]
That makes sense. I am okay with this as long as we can offer clear instructions and help for making self-signed certificates for both Linux and Windows. Overall help in this area is lacking online. In most cases, users trialling Telegraf for the first time against their OPC UA server won't want to be road blocked by tricky certificate creation. A lot of OPC UA clients do offer self-signed certification for testing convenience. :)

**Comment by Jayclifford345** [comment]
I also think based on the OPC UA foundation spec they recommend the auto generation of a self signed cert on installation of OPC UA clients: https://reference.opcfoundation.org/v104/Core/docs/Part2/8/#:~:text=From%20a%20developer%20point%20of%20view%2C%20it%20is%20a%20best%20practice%2C%20if%20your%20OPC%20UA%20Application%20supports%20Certificates%2C%20that%20it%20automatically%20provides%20a%20self%2Dsigned%20Application%20Instance%20Certificate%20on%20installation.

**Comment by srebhan** [comment]
Well we could do this if we can agree on a _fixed_ location and generate a cert if there is none. However, how do you determine that location without interfering with the usual certs?

**Comment by Jayclifford345** [comment]
This is why I thought if the user leaves the two config paths empty this would auto-generate the certificates as it has done but instead of generating a new one on each restart, it would just reuse the originally generated one.

 We would then put in the documentation the location of where the auto-generated cert is kept. I believe @R290 said for Linux this is `/tmp/ssc#` . Users can delete this file if they wish Telegraf to generate a new cert. 
 
 Personally, it would be great if there was just a directory in the Telegraf folder called certificates where all auto-generated certificates go. Though I believe this could be a lot of breaking changes.

**Comment by Jayclifford345** [comment]
Do you think it's worth just discussing the next maintainers meeting? :)

**Comment by MrWinkyTysilio** [comment]
Hello Everyone, 
I still have this issue using telegraf (1.27.3) with Beckhoff OPC UA server.
I already have tried to used certificates generated in /tmp/ssc#: I copy pasted it into /etc/telegraf/ and I update the config file.
But even with this telegraf is creating a new certificate at each connection.
It is working with "None" but I can not use it like that for security reasons.

**Comment by skartikey** [comment]
@srebhan To summarise the issue: Telegraf currently generates self-signed client certificates in temporary directories (using `os.CreateTemp()`), which are cleaned up on restart. As a result, OPC UA servers see Telegraf as a new client each time, requiring manual re-authorization.

I’d like to propose adding a new configuration option, `certificate_storage_path`, that allows users to specify a persistent directory for storing auto-generated client certificates. When configured, Telegraf would generate certificates once and reuse them across restarts. Certificates would be named based on the endpoint to support multiple OPC UA connections with unique certificates. This preserves backward compatibility while solving the persistent identity issue.

**Example configuration:**

```toml
[[inputs.opcua]]
  endpoint = "opc.tcp://localhost:4840"
  security_policy = "Basic256Sha256"
  security_mode = "SignAndEncrypt"

  ## Directory to store auto-generated client certificates (optional)
  ## If not specified, certificates are generated in a temporary directory (current behavior)
  ## If specified, certificates are generated once and reused across restarts
  certificate_storage_path = "/etc/telegraf/opcua/certs"

  ## Alternatively, continue using explicit certificate paths for full control
  # certificate = "/etc/telegraf/cert.pem"
  # private_key = "/etc/telegraf/key.pem"
```

This proposal gives users flexibility to choose between:

* Ephemeral certificates (current behavior)
* Persistent auto-generated certificates (new option)
* Explicit certificate management (existing capability)

What are your thoughts on this approach?

**Comment by srebhan** [comment]
@skartikey to be honest the biggest question is, why don't users just generate a certificate manually if they want persistence? This is the fundamental question...

What we could also do is to store the certificate using [state-persistence](https://github.com/influxdata/telegraf/blob/master/docs/specs/tsd-003-state-persistence.md) which already exists...

Don't get me wrong, I like your idea as it is a clean design but I wonder what we gain for adding more code that you can't have today...

**Comment by skartikey** [comment]
@srebhan You're right to question this - I think there's been some confusion on my part.

PR #17825 solved issue #17608 (trusting server self-signed certificates via `remote_certificate`), which is a different problem than what this issue describes.

  This issue is specifically about **client** certificate persistence - the certificates that Telegraf uses to authenticate itself to the OPC UA server. When these regenerate on restart, users must re-approve them server-side.

  The `remote_certificate` feature addresses server -> client trust.
  This issue is about client -> server trust persistence.

  Given that:
  1. Manual certificate generation already works (documented workaround)
  2. The issue only affects users wanting auto-generated client certs
  3. State-persistence might be overkill for configuration

  I agree this might not warrant additional implementation. Should we close this as "won't fix" and improve documentation instead?

**Comment by srebhan** [comment]
@skartikey how much code are we talking about?

**Comment by skartikey** [comment]
@srebhan As discussed on the call, I’ve implemented the changes and it’s ready for the first review.

The certificate behavior now depends on the configuration:

* **Both `certificate` and `private_key` empty** → Generate temporary certificates in the temp directory (current behavior, backward compatible)
* **Both `certificate` and `private_key` specified but files don’t exist** → Generate persistent certificates at the specified paths (**new**, resolves restart issue)
* **Both `certificate` and `private_key` specified and files exist** → Use existing certificates (current behavior)
